### PR TITLE
refactor: SendEthereumClaims with a more idiomatic approach

### LINKED ISSUE #19: Refactor SendEthereumClaims to make it more readable
## Summary

The function SendEthereumClaims (orchestrator/cosmos/broadcast.go) goes through all the scanned Eth events and relays them over to Umee. Right now it works perfectly but has poor readability.

## Proposal

The proposed solution adds some overhead but offers way more readability. Instead of tracking 4 different lists of events, let's add them to a single list using a wrapper (SortableEvent) and iterate through them.

```go
// SortableEvent exists with the only purpose to make a nicer sortable slice
	type SortableEvent struct {
		EventNonce         uint64
		DepositEvent       *wrappers.PeggySendToCosmosEvent
		WithdrawEvent      *wrappers.PeggyTransactionBatchExecutedEvent
		ValsetUpdateEvent  *wrappers.PeggyValsetUpdatedEvent
		ERC20DeployedEvent *wrappers.PeggyERC20DeployedEvent
	}
	allevents := []SortableEvent{}
	// add all events to allevents
	// sort allevents by event nonce
	// remove all events which nonce is equal or lower than lastClaimEvent
	for _, event := range allevents {
		if event.DepositEvent != nil {
			if err := s.sendDepositClaims(ctx, event.DepositEvent); err != nil {
				log.WithError(err).Errorln("broadcasting MsgDepositClaim failed")
				return err
			}
		}

		if event.WithdrawEvent != nil {
			if err := s.sendWithdrawClaims(ctx, event.WithdrawEvent); err != nil {
				log.WithError(err).Errorln("broadcasting MsgWithdrawClaim failed")
				return err
			}
		}

		// do the same for valsets
		// do the same for erc20 deployed
	}
```
----

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [x] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by facundomedica**
This is meant to close #19

### CODE REVIEW

File: orchestrator/cosmos/broadcast.go [hunk 1]
```diff
@@ -371,47 +373,97 @@ func (s *peggyBroadcastClient) SendEthereumClaims(
...
-
-	// Individual arrays (deposits, withdraws, valsetUpdates) are sorted.
-	// Broadcast claim events sequentially starting with eventNonce = lastClaimEvent + 1.
-	for count < totalClaimEvents {
-		//nolint:gocritic // (we are already working on refactoring this!)
-		if i < len(deposits) && deposits[i].EventNonce.Uint64() == lastClaimEvent+1 {
-			// send deposit
-			if err := s.sendDepositClaims(deposits[i]); err != nil {
+	// SortableEvent exists with the only purpose to make a nicer sortable slice
+	type SortableEvent struct {
```
Comment (alexanderbez) [inline orchestrator/cosmos/broadcast.go:379]: nit: can we define this type outside the method?

File: orchestrator/cosmos/broadcast.go [hunk 2]
```diff
@@ -371,47 +373,97 @@ func (s *peggyBroadcastClient) SendEthereumClaims(
...
+	}
+
+	// Use SliceStable so we always get the same order
+	sort.SliceStable(allevents, func(i, j int) bool {
+		return allevents[i].EventNonce < allevents[j].EventNonce
+	})
+
+	// iterate through events and send them sequentially
+	for _, ev := range allevents {
+		// if the event nonce isn't sequential, we break from this loop
```
Comment (alexanderbez) [inline orchestrator/cosmos/broadcast.go:433]:
```suggestion
		// If the event nonce isn't sequential, we break from this loop
```

File: orchestrator/cosmos/broadcast.go [hunk 3]
```diff
@@ -371,47 +373,97 @@ func (s *peggyBroadcastClient) SendEthereumClaims(
...
+		// given that the events are sorted, this should never happen
```
Comment (alexanderbez) [inline orchestrator/cosmos/broadcast.go:434]:
```suggestion
		// given that the events are sorted, this should never happen.
```

File: orchestrator/cosmos/broadcast_test.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockCosmos := mocks.NewMockCosmosClient(mockCtrl)
+
+	mockCosmos.EXPECT().FromAddress().Return(sdk.AccAddress{}).AnyTimes()
+
+	biggerNonceMatcher := HasBiggerNonce(0)
+
+	mockCosmos.EXPECT().SyncBroadcastMsg(biggerNonceMatcher).Return(&sdk.TxResponse{}, nil).Times(8)
```
Comment (alexanderbez) [inline orchestrator/cosmos/broadcast_test.go:77]:
```suggestion
	mockCosmos := mocks.NewMockCosmosClient(mockCtrl)
	mockCosmos.EXPECT().FromAddress().Return(sdk.AccAddress{}).AnyTimes()

	biggerNonceMatcher := HasBiggerNonce(0)
	mockCosmos.EXPECT().SyncBroadcastMsg(biggerNonceMatcher).Return(&sdk.TxResponse{}, nil).Times(8)
```

File: orchestrator/cosmos/broadcast_test.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockCosmos := mocks.NewMockCosmosClient(mockCtrl)
+
+	mockCosmos.EXPECT().FromAddress().Return(sdk.AccAddress{}).AnyTimes()
+
+	biggerNonceMatcher := HasBiggerNonce(0)
+
+	mockCosmos.EXPECT().SyncBroadcastMsg(biggerNonceMatcher).Return(&sdk.TxResponse{}, nil).Times(7)
```
Comment (alexanderbez) [inline orchestrator/cosmos/broadcast_test.go:151]:
```suggestion
func TestSendEthereumClaimsIgnoreNonSequentialNonces(t *testing.T) {
	mockCtrl := gomock.NewController(t)
	defer mockCtrl.Finish()

	mockCosmos := mocks.NewMockCosmosClient(mockCtrl)
	mockCosmos.EXPECT().FromAddress().Return(sdk.AccAddress{}).AnyTimes()

	biggerNonceMatcher := HasBiggerNonce(0)
	mockCosmos.EXPECT().SyncBroadcastMsg(biggerNonceMatcher).Return(&sdk.TxResponse{}, nil).Times(7)
```

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Looks great @facundomedica. Since we're introducing mocks, I would highly encourage we add a make target to generate them. Otherwise, other developers have no idea how to update mocks.

File: Makefile [hunk 6]
```diff
@@ -52,7 +52,13 @@ lint:
 	@echo "--> Running linter"
 	@go run github.com/golangci/golangci-lint/cmd/golangci-lint run --timeout=10m
 
-.PHONY: test-integration lint
+mocks:
+	@echo "--> Generating mocks"
+	@go run github.com/golang/mock/mockgen -destination=mocks/cosmos.go \
```
Comment (alexanderbez) [inline Makefile:57]: `github.com/golang/mock/mockgen` should be added to the `/tools/tools.go` package 👍

File: go.mod [hunk 7]
```diff
@@ -37,12 +37,17 @@ require (
...
+	golang.org/x/crypto v0.0.0-20210817164053-32db794688a5 // indirect
 	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
 	google.golang.org/grpc v1.41.0
 	gopkg.in/yaml.v2 v2.4.0
 )
 
+require (
+	github.com/golang/mock v1.6.0
+	golang.org/x/term v0.0.0-20210220032956-6a3ed077a48d
+)
```
Comment (alexanderbez) [inline go.mod:49]: I would put these in the require above and run `go mod tidy -compat=1.17`

**Review by alexanderbez — APPROVED** [review]

File: go.mod (hunk 7 above)
Comment (facundomedica) [reply go.mod]: OK, ran `go mod tidy -compat=1.17 `

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**